### PR TITLE
feat(tiering): add max file size limit

### LINKED ISSUE #2207: add max_file_size flag to limit the size of SSD writes
we will split the max_file_size equaly between shards so that each shard will be able to write to file max_file_size/number_of_shards bytes

### PR DESCRIPTION
**Pull request by adiholden**
fixes #2207

### CODE REVIEW

File: src/server/engine_shard_set.cc [hunk 1]
```diff
@@ -29,6 +33,11 @@ ABSL_FLAG(string, spill_file_prefix, "",
           " associated with tiered storage. E.g,"
           "spill_file_prefix=/path/to/file-prefix");
 
+ABSL_FLAG(dfly::MemoryBytesFlag, max_file_size, dfly::MemoryBytesFlag{},
```
Comment (romange) [inline src/server/engine_shard_set.cc:36]: I suggest prefixing this flag with a feature name, aka `tiered_file_max_size`.
or `spill_file_max_size` for consistency though I think we won't use `spill file` externally so it probably not a good name,

File: src/server/engine_shard_set.cc [hunk 2]
```diff
@@ -787,14 +797,52 @@ auto EngineShard::AnalyzeTxQueue() -> TxQueueInfo {
 
  */
 
+uint64_t GetDiskLimit() {
```
Comment (romange) [inline src/server/engine_shard_set.cc:800]: `GetDiskLimit` -> `GetFsLimit`

File: src/server/engine_shard_set.cc [hunk 3]
```diff
@@ -29,6 +33,11 @@ ABSL_FLAG(string, spill_file_prefix, "",
           " associated with tiered storage. E.g,"
           "spill_file_prefix=/path/to/file-prefix");
 
+ABSL_FLAG(dfly::MemoryBytesFlag, max_file_size, dfly::MemoryBytesFlag{},
+          "Limit on maximum file size that is used by the database for tiered storage. "
+          "0 - means the program will automatically determine its maximum file size. "
```
Comment (romange) [inline src/server/engine_shard_set.cc:38]: is it per shard or global?

Comment (adiholden) [reply src/server/engine_shard_set.cc:38]: global

File: src/server/engine_shard_set.cc [hunk 4]
```diff
@@ -787,14 +797,52 @@ auto EngineShard::AnalyzeTxQueue() -> TxQueueInfo {
...
+      if (max_file_size_limit < max_file_size) {
+        LOG(WARNING) << "Got max file size " << HumanReadableNumBytes(max_file_size)
+                     << ", however only " << HumanReadableNumBytes(max_file_size_limit)
+                     << " disk space was found.";
+      }
+    }
+    max_shard_file_size = max_file_size / shard_queue_.size();
+    if (max_shard_file_size < 256_MB) {
+      LOG(ERROR) << "Max tiering file size is too small. Setting: "
+                 << HumanReadableNumBytes(max_file_size) << " Requeried at least "
```
Comment (romange) [inline src/server/engine_shard_set.cc:836]: Requeried -> Required

**Comment by adiholden** [comment]
@romange ping

**Review by romange — APPROVED** [review]

**Pull request merged**